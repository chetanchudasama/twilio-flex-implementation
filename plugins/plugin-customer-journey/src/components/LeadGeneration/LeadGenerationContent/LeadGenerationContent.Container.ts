import { connect } from "react-redux";
import { Dispatch } from "redux";
import { CustomerDetailModel } from "@common/components";

import { AppState, namespace } from "../../../states";
import { LeadGenerationContent } from "./LeadGenerationContent";
import { StateToProps, DispatchToProps } from "./LeadGenerationContent.Props";

const mapStateToProps = (state: AppState): StateToProps => {
  const customer = Object.assign(
    new CustomerDetailModel(),
    state[namespace].customer.customer
  );
  return {
    applicationId: customer.applicationId ?? 0,
    customerName: customer.fullName,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>): DispatchToProps => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LeadGenerationContent);
