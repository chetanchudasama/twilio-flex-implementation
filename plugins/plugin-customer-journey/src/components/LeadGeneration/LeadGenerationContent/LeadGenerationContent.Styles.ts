import styled from "react-emotion";
import {
  darkGrey,
  lightGrey,
  white,
  scrollbarColor,
  black,
  mediumGrey,
  leadGenBorderColor,
} from "@common/components";

export const LeadGenerationContentStyles = styled("div")({
  overflowY: "hidden",
  overflowX: "hidden",
  width: "100%",
  height: "100%",
  ".lead-gen-container": {
    fontSize: 14,
    padding: "16px 24px",
    height: "100%",
    width: "100%",
    background: lightGrey,
    marginLeft: 0,
    marginTop: 0,
    "&::-webkit-scrollbar": {
      width: "0.6em",
    },
    "&::-webkit-scrollbar-track": {
      background: lightGrey,
    },
    "&::-webkit-scrollbar-thumb": {
      background: darkGrey,
    },
    "&::-webkit-scrollbar-thumb:hover": {
      background: scrollbarColor,
    },
  },
  ".hide-scrollbar": {
    overflowY: "hidden",
  },
  ".show-scrollbar": {
    overflowY: "auto",
  },
  ".white-background": {
    background: white,
    margin: "12px 0",
    paddingTop: "0 !important",
    paddingBottom: "0 !important",
    border: `1px solid ${leadGenBorderColor}`,
    borderRadius: 4,
    boxShadow: "0px 1px 2px rgba(44, 43, 53, 0.04)",
  },
  ".customer-name": {
    paddingLeft: "2px !important",
    fontSize: 25,
    fontWeight: 600,
    color: darkGrey,
    display: "flex",
    wordBreak: "break-word",
  },
  ".step-container": {
    display: "flex",
  },
  ".step-name": {
    marginLeft: 16,
    marginRight: 4,
    marginTop: 22,
    fontSize: 20,
    fontWeight: 600,
    color: black,
    opacity: 0.5,
  },
  ".divider-left": {
    borderLeft: `1px solid ${mediumGrey}`,
    height: 40,
    margin: "0 16px",
  },
  ".menu-container": {
    display: "flex",
    justifyContent: "flex-end",
    paddingRight: "0px !important",
  },
  ".menu-container .icon-btn": {
    minHeight: 40,
    background: white,
  },
});
