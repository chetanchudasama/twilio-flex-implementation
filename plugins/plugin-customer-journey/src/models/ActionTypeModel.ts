export class ActionTypeModel {
  public id: number;

  public name: string;

  constructor() {
    this.id = -1;
    this.name = "";
  }
}
