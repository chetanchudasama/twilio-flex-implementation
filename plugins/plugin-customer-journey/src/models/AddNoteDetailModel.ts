export class AddNoteDetailModel {
  public applicationId: number;

  public content: string;

  public isImportant: boolean;

  public sendToBroker: boolean;

  constructor() {
    this.applicationId = -1;
    this.content = "";
    this.isImportant = false;
    this.sendToBroker = false;
  }
}
