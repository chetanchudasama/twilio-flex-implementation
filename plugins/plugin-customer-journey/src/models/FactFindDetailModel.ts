import { ItemModel } from "@common/components";
import { TimeForPurchaseItemModel } from "./TimeForPurchaseModel";

export class FactFindDetailModel {
  public borrowAmount: number;

  public hasCustomerAlreadyFoundCar: boolean | null;

  public makeId: number;

  public modelId: number;

  public vrm: string;

  public vehicleType: ItemModel;

  public timeForPurchase: TimeForPurchaseItemModel | null;

  public note: string;

  constructor() {
    this.borrowAmount = 0;
    this.hasCustomerAlreadyFoundCar = null;
    this.makeId = -1;
    this.modelId = -1;
    this.vrm = "";
    this.vehicleType = new ItemModel();
    this.timeForPurchase = new TimeForPurchaseItemModel();
    this.note = "";
  }
}
