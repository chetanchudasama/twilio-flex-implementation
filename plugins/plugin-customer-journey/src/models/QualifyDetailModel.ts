import { AgentGuideContentModel } from "./AgentGuideContentModel";
import { ReasonForCarPurchaseItemModel } from "./ReasonForCarPurchaseItemModel";
import { TimeForPurchaseItemModel } from "./TimeForPurchaseModel";
import { WhereDidYouHearItemModel } from "./WhereDidYouHearItemModel";

export class QualifyDetailModel {
  public drivingLicenseConfirmed: boolean | null;

  public addressConfirmed: boolean | null;

  public incomeConfirmed: boolean | null;

  public placeOfWorkConfirmed: boolean | null;

  public reasonForPurchase: ReasonForCarPurchaseItemModel | null;

  public timeForPurchase: TimeForPurchaseItemModel | null;

  public whereDidYouHear: WhereDidYouHearItemModel | null;

  public agentGuideContent: AgentGuideContentModel;

  public hasCustomerAlreadyFoundCar: boolean | null;

  constructor() {
    this.drivingLicenseConfirmed = null;
    this.addressConfirmed = null;
    this.incomeConfirmed = null;
    this.placeOfWorkConfirmed = null;
    this.reasonForPurchase = new ReasonForCarPurchaseItemModel();
    this.timeForPurchase = new TimeForPurchaseItemModel();
    this.whereDidYouHear = new WhereDidYouHearItemModel();
    this.agentGuideContent = new AgentGuideContentModel();
    this.hasCustomerAlreadyFoundCar = null;
  }
}
