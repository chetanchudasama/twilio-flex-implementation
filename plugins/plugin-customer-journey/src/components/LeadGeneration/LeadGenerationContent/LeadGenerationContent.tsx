import React, { useState } from "react";
import { Grid } from "@material-ui/core";
import { CustomerBannerMenu, CustomerDetailType } from "@common/components";
import { LeadGenerationContentStyles } from "./LeadGenerationContent.Styles";
import { Allocation } from "../Allocation/Allocation";
import { UpdateApplication } from "../UpdateApplication/UpdateApplication";
import FactFind from "../FactFind/FactFind.Container";
import QualifyDetail from "../QualifyDetail/QualifyDetail.Container";
import CustomerName from "../CustomerName/CustomerName.Container";
import DPAFCA from "../DPAFCA/DPAFCA.Container";
import LenderDetail from "../LenderDetail/LenderDetail.Container";
import { CampaignDetails } from "../CampaignDetails/CampaignDetails";
import { RecentNotes } from "../RecentNotes/RecentNotes";
import ContactOutcome from "../ContactOutcome/ContactOutcome.Container";
import { LeadGenerationContentProps } from "./LeadGenerationContent.Props";
import CustomerDetailActionContainer from "../../CustomerDetailAction/CustomerDetailActionContent.Container";

export const LeadGenerationContent: React.FC<LeadGenerationContentProps> = ({
  applicationId,
  customerName,
  qualifyingDetail,
  updateQualifyingDetails,
  updateApplication,
  factFindDetail,
  updateFactFindDetail,
  recentNotes,
  updateCustomerPreferredName,
  setCallbackDetail,
  getRankedAgents,
  handleContactOutcomeClick,
}) => {
  const [customerDetailType, setCustomerDetailType] =
    useState<CustomerDetailType | null>(null);

  const updateCustomerDetail = (type: CustomerDetailType) => {
    setCustomerDetailType(type);
  };

  return (
    <LeadGenerationContentStyles>
      <Grid
        container
        spacing={16}
        className={`lead-gen-container ${
          customerDetailType ? "hide-scrollbar" : "show-scrollbar"
        }`}
      >
        <Grid item xs={12}>
          <Grid container spacing={16} className="customer-name-lg">
            <Grid item xs={5} className="customer-name">
              {`${customerName} (#${applicationId})`}
            </Grid>
            {!customerDetailType && (
              <Grid item xs={7} className="menu-container">
                <ContactOutcome
                  setCallbackDetail={setCallbackDetail}
                  handleContactOutcomeClick={handleContactOutcomeClick}
                />
                <span className="divider-left" />
                <CustomerBannerMenu
                  updateCustomerDetail={updateCustomerDetail}
                  isFromLeadGenerationScreen
                />
              </Grid>
            )}
          </Grid>
        </Grid>
        {customerDetailType ? (
          <CustomerDetailActionContainer
            backToWizardStepper={() => {
              setCustomerDetailType(null);
            }}
            customerDetailType={customerDetailType}
            isFromLeadGenerationScreen
          />
        ) : (
          <>
            <Grid item xs={12}>
              <Grid container spacing={16}>
                <Grid item xs={4}>
                  <CampaignDetails />
                </Grid>
                <Grid item xs={8}>
                  <RecentNotes recentNotes={recentNotes} />
                </Grid>
              </Grid>
            </Grid>
            <Grid item xs={12} className="white-background">
              <LenderDetail />
            </Grid>
            <Grid item xs={12} className="white-background step-container">
              <span className="step-name">01</span>
              <DPAFCA />
            </Grid>
            <Grid item xs={12} className="white-background">
              <CustomerName
                updateCustomerPreferredName={updateCustomerPreferredName}
              />
            </Grid>
            <Grid item xs={12} className="white-background step-container">
              <span className="step-name">02</span>
              <QualifyDetail
                qualifyingDetail={qualifyingDetail}
                updateQualifyingDetails={updateQualifyingDetails}
              />
            </Grid>
            <Grid item xs={12} className="white-background step-container">
              <span className="step-name">03</span>
              <FactFind
                factFindDetail={factFindDetail}
                updateFactFindDetail={updateFactFindDetail}
              />
            </Grid>
            <Grid item xs={12} className="white-background step-container">
              <span className="step-name">04</span>
              <UpdateApplication updateApplication={updateApplication} />
            </Grid>
            <Grid item xs={12} className="white-background step-container">
              <span className="step-name">05</span>
              <Allocation getRankedAgents={getRankedAgents} />
            </Grid>
          </>
        )}
      </Grid>
    </LeadGenerationContentStyles>
  );
};
