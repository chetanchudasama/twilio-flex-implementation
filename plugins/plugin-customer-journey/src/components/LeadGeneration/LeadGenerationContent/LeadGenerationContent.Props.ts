import { ItemModel } from "@common/components";

import { QualifyDetailModel } from "../../../models/QualifyDetailModel";
import { FactFindDetailModel } from "../../../models/FactFindDetailModel";
import { TimeForPurchaseItemModel } from "../../../models/TimeForPurchaseModel";
import { NoteModel } from "../../../models/NoteModel";
import { TaskDecisionModel } from "../../../models/TaskDecisionModel";

export interface StateToProps {
  applicationId: number;
  customerName: string;
}

export interface DispatchToProps {}

export interface OwnProps {
  qualifyingDetail: QualifyDetailModel;
  updateQualifyingDetails: (
    key: keyof QualifyDetailModel,
    value: boolean | null
  ) => void;
  updateApplication: () => void;
  factFindDetail: FactFindDetailModel;
  updateFactFindDetail: (
    key: string,
    value:
      | string
      | number
      | boolean
      | TimeForPurchaseItemModel
      | ItemModel
      | null
  ) => void;
  recentNotes: NoteModel[];
  updateCustomerPreferredName: (preferredName: string) => void;
  setCallbackDetail: (callbackDate: Date | null) => void;
  getRankedAgents: () => void;
  handleContactOutcomeClick: (
    taskDecision: TaskDecisionModel,
    callbackBooked?: Date
  ) => void;
}

export type LeadGenerationContentProps = StateToProps &
  DispatchToProps &
  OwnProps;
