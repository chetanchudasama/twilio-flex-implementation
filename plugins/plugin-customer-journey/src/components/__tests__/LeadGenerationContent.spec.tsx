import { create } from "react-test-renderer";
import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";

import { CustomerBannerMenu, CustomerDetailType } from "@common/components";
import { QualifyDetailModel } from "../../models/QualifyDetailModel";
import { LeadGenerationContent } from "../LeadGeneration/LeadGenerationContent/LeadGenerationContent";
import { LeadGenerationContentProps } from "../LeadGeneration/LeadGenerationContent/LeadGenerationContent.Props";
import { FactFindDetailModel } from "../../models/FactFindDetailModel";
import CustomerDetailActionContainer from "../CustomerDetailAction/CustomerDetailActionContent.Container";

const props: LeadGenerationContentProps = {
  applicationId: 1,
  customerName: "Jon Smith",
  qualifyingDetail: new QualifyDetailModel(),
  updateQualifyingDetails: jest.fn(),
  updateApplication: jest.fn(),
  factFindDetail: new FactFindDetailModel(),
  updateFactFindDetail: jest.fn(),
  recentNotes: [],
  updateCustomerPreferredName: jest.fn(),
  setCallbackDetail: jest.fn(),
  getRankedAgents: jest.fn(),
  handleContactOutcomeClick: jest.fn(),
};

jest.mock("@common/components", () => {
  const components = jest.requireActual("@common/components");
  return {
    ...components,
    CustomerBannerMenu: jest.fn(({ children }) => {
      return <div>{children}</div>;
    }),
  };
});

jest.mock("../LeadGeneration/ContactOutcome/ContactOutcome.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Contact Outcome</div>;
    },
  };
});

jest.mock("../LeadGeneration/CampaignDetails/CampaignDetails", () => {
  return {
    CampaignDetails: () => {
      return <div>Campaign Details</div>;
    },
  };
});

jest.mock("../LeadGeneration/Allocation/Allocation", () => {
  return {
    Allocation: () => {
      return <div>Allocation</div>;
    },
  };
});

jest.mock("../LeadGeneration/CustomerName/CustomerName.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Customer Name</div>;
    },
  };
});

jest.mock("../LeadGeneration/DPAFCA/DPAFCA.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>DPA FCA</div>;
    },
  };
});

jest.mock("../LeadGeneration/FactFind/FactFind.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Fact Find</div>;
    },
  };
});

jest.mock("../LeadGeneration/LenderDetail/LenderDetail.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Lender Detail</div>;
    },
  };
});

jest.mock("../LeadGeneration/QualifyDetail/QualifyDetail.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Qualify Detail</div>;
    },
  };
});

jest.mock("../LeadGeneration/RecentNotes/RecentNotes", () => {
  return {
    RecentNotes: () => {
      return <div>Recent Notes</div>;
    },
  };
});

jest.mock("../LeadGeneration/UpdateApplication/UpdateApplication", () => {
  return {
    UpdateApplication: () => {
      return <div>Update Application</div>;
    },
  };
});

jest.mock(
  "../CustomerDetailAction/CustomerDetailActionContent.Container",
  () => {
    return {
      __esModule: true,
      default: jest.fn(({ children }) => {
        return <div>{children}</div>;
      }),
    };
  }
);

describe("LeadGenerationContent", () => {
  it("renders correctly", () => {
    const wrapper = create(<LeadGenerationContent {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should render CustomerDetailActionContainer component, when updateCustomerDetail method is called from CustomerBannerMenu", () => {
    const wrapper = mount(<LeadGenerationContent {...props} />);

    act(() => {
      wrapper
        .find(CustomerBannerMenu)
        .props()
        .updateCustomerDetail(CustomerDetailType.CustomerDetail);
    });
    wrapper.update();

    expect(wrapper.find(CustomerDetailActionContainer).exists()).toBe(true);
    expect(wrapper.find(".hide-scrollbar").exists()).toBe(true);
  });

  it("should hide CustomerDetailActionContainer component, when backToWizardStepper method is called", () => {
    const wrapper = mount(<LeadGenerationContent {...props} />);

    act(() => {
      wrapper
        .find(CustomerBannerMenu)
        .props()
        .updateCustomerDetail(CustomerDetailType.CustomerDetail);
    });
    wrapper.update();

    act(() => {
      wrapper.find(CustomerDetailActionContainer).props().backToWizardStepper();
    });
    wrapper.update();

    expect(wrapper.find(CustomerDetailActionContainer).exists()).toBe(false);
    expect(wrapper.find(".show-scrollbar").exists()).toBe(true);
  });
});
