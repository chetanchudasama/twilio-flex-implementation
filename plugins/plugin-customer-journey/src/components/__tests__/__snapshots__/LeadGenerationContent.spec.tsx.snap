// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeadGenerationContent renders correctly 1`] = `
<div
  className="css-do6ydh"
>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24 lead-gen-container show-scrollbar"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
    >
      <div
        className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24 customer-name-lg"
      >
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-5-34 customer-name"
        >
          Jon Smith (#1)
        </div>
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-7-36 menu-container"
        >
          <div>
            Contact Outcome
          </div>
          <span
            className="divider-left"
          />
          <div />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
    >
      <div
        className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24"
      >
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-4-33"
        >
          <div>
            Campaign Details
          </div>
        </div>
        <div
          className="MuiGrid-item-2 MuiGrid-grid-xs-8-37"
        >
          <div>
            Recent Notes
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background"
    >
      <div>
        Lender Detail
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background step-container"
    >
      <span
        className="step-name"
      >
        01
      </span>
      <div>
        DPA FCA
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background"
    >
      <div>
        Customer Name
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background step-container"
    >
      <span
        className="step-name"
      >
        02
      </span>
      <div>
        Qualify Detail
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background step-container"
    >
      <span
        className="step-name"
      >
        03
      </span>
      <div>
        Fact Find
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background step-container"
    >
      <span
        className="step-name"
      >
        04
      </span>
      <div>
        Update Application
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 white-background step-container"
    >
      <span
        className="step-name"
      >
        05
      </span>
      <div>
        Allocation
      </div>
    </div>
  </div>
</div>
`;
